cpp/include/cudf/groupby.hpp

```diff
@@ -222,6 +222,42 @@ class groupby {
     host_span<aggregation_request const> requests,
     rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+  /**
+   * @brief Performs grouped shifts for specified values.
+   *
+   * For each group, `i`th element is determined by the `i - offset`th element
+   * of the group. If `i - offset < 0 or >= group_size`, the value is determined by
+   * @p fill_value.
+   *
+   * Example:
+   * @code{.pseudo}
+   * keys:   {1 1 1 1 2 2 2}
+   * values: {3 1 4 7 9 2 5}
+   * offset: 2
+   * fill_value: @
+   * result: {@ @ 3 1 @ @ 9}
+   * -------------------------------------------------
+   * keys:   {1 1 1 1 2 2 2}
+   * values: {3 1 4 7 9 2 5}
+   * offset: -2
+   * fill_value: -1
+   * result: {4 7 -1 -1 5 -1 -1}
+   * @endcode
+   *
+   * @param values Column to be shifted
+   * @param offset The off set by which to shift the input
+   * @param fill_value Fill value for indeterminable outputs
+   * @param mr Device memory resource used to allocate the returned table and columns' device memory
+   * @return Pair containing the table with each group's key and the column shifted
+   *
+   * @throws cudf::logic_error if @p fill_value dtype does not match @p input dtype
+   */
+  std::pair<std::unique_ptr<table>, std::unique_ptr<column>> shift(
```

Comment:
Is there a need for an API that allows shifting each group by a different amount? E.g. it takes a column of shift amounts rather than a single offset? Just curious.

Reply:
Giving some thoughts plus picking @shwina 's brain, there isn't a use case that pops up immediately to us. Pandas groupby shift does not take 1 shift offset for each group:
```python
df = pd.DataFrame({'a': [1, 1, 2, 2, 3, 4], 'b': [2, 3, 4, 5, 6, 7]})
df.groupby('a').shift([1, 1, 0, 0])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-087bae55558a> in <module>()
      1 df = pd.DataFrame({'a': [1, 1, 2, 2, 3, 4], 'b': [2, 3, 4, 5, 6, 7]})
----> 2 df.groupby('a').shift([1, 1, 0, 0])

1 frames
/usr/local/lib/python3.7/dist-packages/pandas/core/groupby/groupby.py in _get_cythonized_result(self, how, cython_dtype, aggregate, numeric_only, needs_counts, needs_values, needs_2d, min_count, needs_mask, needs_ngroups, result_is_index, pre_processing, post_processing, **kwargs)
   2507                 func = partial(func, ngroups)
   2508 
-> 2509             func(**kwargs)  # Call func to modify indexer values in place
   2510 
   2511             if needs_2d:

pandas/_libs/groupby.pyx in pandas._libs.groupby.group_shift_indexer()

TypeError: an integer is required
```

For its underlying `segmented_shift` usage is a bit nuanced. For LIST columns, in theory you can do this in Pandas:
```python
>>> df = pd.DataFrame({'offset': [1, -2], 'lists': [[1, 2, 3], [4, 5, 6]]})
>>> fill = None
>>> def shift_row(x):
...   offset, l = x['offset'], x['lists']
...   if offset > 0:
...     return [fill]*offset+l[:-offset]
...   else:
...     return l[-offset:]+[fill]*(-offset)
>>> df.apply(shift_row , axis=1)
0       [None, 1, 2]
1    [6, None, None]
dtype: object
```
I guess we could build an overload for `segmented_shift` with a column of offsets if requested.